File: common/src/main/java/io/github/notenoughupdates/moulconfig/xml/loaders/SpacerLoader.java
package io.github.notenoughupdates.moulconfig.xml.loaders;

import io.github.notenoughupdates.moulconfig.gui.component.SpacerComponent;
import io.github.notenoughupdates.moulconfig.observer.GetSetter;
import io.github.notenoughupdates.moulconfig.xml.ChildCount;
import io.github.notenoughupdates.moulconfig.xml.XMLContext;
import io.github.notenoughupdates.moulconfig.xml.XMLGuiLoader;
import io.github.notenoughupdates.moulconfig.xml.XMLUniverse;
import org.w3c.dom.Element;

import javax.xml.namespace.QName;
import java.util.LinkedHashMap;
import java.util.Map;

public class SpacerLoader implements XMLGuiLoader.Basic<SpacerComponent> {
    @Override public SpacerComponent createInstance(XMLContext<?> context, Element element) {
        GetSetter<Integer> width = context.getPropertyFromAttribute(element, new QName("width"), Integer.class);
        GetSetter<Integer> height = context.getPropertyFromAttribute(element, new QName("height"), Integer.class);
        return new SpacerComponent(width != null ? width : GetSetter.constant(0), height != null ? height : GetSetter.constant(0));
    }
    @Override public QName getName() { return XMLUniverse.qName("Spacer"); }
    @Override public ChildCount getChildCount() { return ChildCount.NONE; }
    @Override public Map<String, Boolean> getAttributeNames() {
        Map<String, Boolean> map = new LinkedHashMap<>();
        map.put("width", false);
        map.put("height", false);
        return map;
    }
}

File: common/src/main/java/io/github/notenoughupdates/moulconfig/xml/loaders/SpacerLoader.kt
package io.github.notenoughupdates.moulconfig.xml.loaders

import io.github.notenoughupdates.moulconfig.gui.component.SpacerComponent
import io.github.notenoughupdates.moulconfig.observer.GetSetter
import io.github.notenoughupdates.moulconfig.xml.ChildCount
import io.github.notenoughupdates.moulconfig.xml.XMLContext
import io.github.notenoughupdates.moulconfig.xml.XMLGuiLoader
import io.github.notenoughupdates.moulconfig.xml.XMLUniverse
import org.w3c.dom.Element
import javax.xml.namespace.QName

class SpacerLoader : XMLGuiLoader<SpacerComponent> {
    override fun createInstance(context: XMLContext<*>, element: Element): SpacerComponent {
        return SpacerComponent(
            context.getPropertyFromAttribute(element, QName("width"), Int::class.java) ?: GetSetter.constant(0),
            context.getPropertyFromAttribute(element, QName("height"), Int::class.java) ?: GetSetter.constant(0),
        )
    }

    override fun getName(): QName {
        return XMLUniverse.qName("Spacer")
    }

    override fun getChildCount(): ChildCount {
        return ChildCount.NONE
    }

    override fun getAttributeNames(): Map<String, Boolean> {
        return mapOf("width" to false, "height" to false)
    }
}